### PR TITLE
TChannelOutResponse: stop sending after connection is closed

### PR DESCRIPTION
**Pull request by jcorbin**
r @kriskowal @Raynos

### CODE REVIEW

File: node/out_response.js [hunk 1]
```diff
@@ -195,6 +195,21 @@ TChannelOutResponse.prototype.sendCallResponseContFrame = function sendCallRespo
 
 TChannelOutResponse.prototype.sendError = function sendError(codeString, message) {
     var self = this;
+
+    // if the request has already error with a similar error, stop
+    if (self.inreq.err &&
+        errors.classify(self.inreq.err) === codeString) {
```
Comment (Raynos) [inline node/out_response.js]: Should logger.info()

File: node/out_response.js [hunk 2]
```diff
@@ -337,6 +352,12 @@ TChannelOutResponse.prototype.send = function send(res1, res2) {
...
     }
 
+    // TODO: may be spam, consider dropping
+    if (self.inreq.connection.closing) {
+        self.logger.info('ignoring outresponse.send on a closed connection', self.extendLogInfo({}));
+        return;
+    }
+
     self.sendCallResponseFrame([self.arg1, res1, res2], true);
     self.emitFinish();
```
Comment (kriskowal) [inline node/out_response.js]: These early returns above preclude this emitFinish(). Do all of these cases imply that emitFinish() has already been sent? Perhaps we should emitFinish() up top if it hasn’t been emitted yet.

Also, callers of sendError may be expecting it to return self consistently.

**Comment by Raynos** [comment]
lgtm.

File: node/out_response.js [hunk 3]
```diff
@@ -195,6 +195,15 @@ TChannelOutResponse.prototype.sendCallResponseContFrame = function sendCallRespo
 
 TChannelOutResponse.prototype.sendError = function sendError(codeString, message) {
     var self = this;
+
+    if (self.inreq.connection.closing) {
+        self.logger.info('ignoring outresponse.sendError on a closed connection', {
+            codeString: codeString,
+            errorMessage: message
+        });
+        return;
```
Comment (kriskowal) [inline node/out_response.js]: should return self here?

Comment (kriskowal) [inline node/out_response.js]: We can get here without being in the done or error state. This should also emit finish before returning.

**Comment by kriskowal** [comment]
:ok:

**Pull request merged**